File: android/app/src/main/kotlin/com/chipbase/chipcore_sdk/flutter/WalletManager.kt
package com.chipcore.sdk.flutter

import android.util.Log
import java.util.Locale

// ─── NativeWalletState ────────────────────────────────────────────────────────

internal class NativeWalletState(private val repository: CardCurrencyRepository) {
    var lastCardId: String? = null
    /** 上次扫卡结果中该卡是否设置了 PIN；用于在签名/发交易前预先弹出原生 PIN 输入框 */
    var lastPinSet: Boolean = false
    private val currencies = LinkedHashMap<String, Messages.CurrencyInfoMessage>()
    private val masters = LinkedHashMap<String, MasterKey>()

    /** 按 id 精确查找 */
    fun findById(id: String): Messages.CurrencyInfoMessage? = currencies[id]

    /** 当前内存中所有不重复的币种（用于持久化） */
    fun allCurrencies(): List<Messages.CurrencyInfoMessage> = currencies.values.distinctBy { it.id }

    fun replaceCurrencies(newCurrencies: List<Messages.CurrencyInfoMessage>) {
        currencies.clear()
        for (currency in newCurrencies) {
            currencies[currency.id] = currency
            currencies[currency.networkId] = currency
        }
    }

    fun mergeCurrencies(newCurrencies: List<Messages.CurrencyInfoMessage>) {
        for (currency in newCurrencies) {
            currencies[currency.id] = currency
            currencies[currency.networkId] = currency
        }
    }

    fun removeCurrencies(coinIds: List<String>) {
        coinIds.forEach(currencies::remove)
    }

    fun findAddress(networkId: String): String? =
        currencies[networkId]?.address ?: currencies.values.firstOrNull { it.networkId == networkId }?.address

    fun updateAddress(networkId: String, address: String) {
        val currency = currencies[networkId] ?: currencies.values.firstOrNull { it.networkId == networkId } ?: return
        val updated = currency.copyWith(address = address)
        currencies[currency.id] = updated
        currencies[currency.networkId] = updated
    }

    fun findPublicKeyHex(keyword: String): String? = currencies.values.firstNotNullOfOrNull { currency ->
        if (
            currency.networkId.contains(keyword, ignoreCase = true) ||
            currency.symbol.contains(keyword, ignoreCase = true) ||
            currency.name.contains(keyword, ignoreCase = true)
        ) {
            currency.publicKey?.toHex()
        } else {
            null
        }
    }

    fun findBySpec(spec: BlockchainSpec): Messages.CurrencyInfoMessage? = currencies.values.firstOrNull { currency ->
        spec.matches(currency.networkId) || spec.matches(currency.symbol) || spec.matches(currency.name)
    }

    fun cacheMasterKey(cardId: String, publicKey: ByteArray, chainCode: ByteArray) {
        masters[cardId] = MasterKey(publicKey, chainCode)
    }

    /** 按 symbol + networkId 查找合约地址（用于 token 发送）*/
    fun findContractAddress(symbol: String, networkId: String): String? {
        val sym = symbol.lowercase()
        val net = networkId.lowercase()
        return currencies.values.firstOrNull { currency ->
            (currency.symbol.lowercase() == sym || currency.id.lowercase() == sym) &&
            currency.networkId.lowercase() == net &&
            !currency.contractAddress.isNullOrEmpty()
        }?.contractAddress
    }

    /** 按 symbol + networkId 查找代币精度（decimals），默认 18 */
    fun findDecimals(symbol: String, networkId: String): Int {
        val sym = symbol.lowercase()
        val net = networkId.lowercase()
        return currencies.values.firstOrNull { currency ->
            (currency.symbol.lowercase() == sym || currency.id.lowercase() == sym) &&
            currency.networkId.lowercase() == net
        }?.decimalCount?.toInt() ?: 18
    }
}

// ─── ChipWalletManager ────────────────────────────────────────────────────────

/**
 * 对应 card_coin 的 WalletManager，管理单条链的地址、公钥及链上操作。
 * 可通过 [WalletManagerRegistry] 按链类型检索。
 */
internal class ChipWalletManager(
    val spec: BlockchainSpec,
    val address: String,
    val publicKey: ByteArray,
    val chainCode: ByteArray,
    val isTest: Boolean,
) {
    /** 最近一次拉取的余额（链原生单位字符串，如 "0.001"），null 表示尚未更新 */
    var balance: String? = null

    fun update() {
        balance = try {
            when (spec) {
                BlockchainSpec.Bitcoin  -> ChainClient.fetchBtcBalance(address, isTest)
                BlockchainSpec.Tron     -> ChainClient.fetchTrxBalance(address)
                BlockchainSpec.Dogecoin -> ChainClient.fetchDogeBalance(address, isTest)
                else                   -> ChainClient.fetchEthBalance(address, isTest)
            }
        } catch (e: Exception) {
            Log.w("ChipCoreNfc", "ChipWalletManager.update failed for ${spec.id}: ${e.message}")
            null
        }
    }

    fun getFees(): List<Messages.FeeResponse> {
        return when (spec) {
            BlockchainSpec.Bitcoin  -> ChainClient.fetchBtcFees(isTest)
            BlockchainSpec.Tron     -> ChainClient.fetchTrxFees()
            BlockchainSpec.Dogecoin -> ChainClient.fetchDogeFees(isTest)
            else                   -> ChainClient.fetchEthFees(isTest)
        }
    }
}

// ─── WalletManagerRegistry ────────────────────────────────────────────────────

/**
 * 对应 card_coin 的 WalletState，管理多条链的 [ChipWalletManager] 实例集合。
 */
internal class WalletManagerRegistry {
    private val managers = LinkedHashMap<String, ChipWalletManager>()

    fun getWalletManager(blockchain: String): ChipWalletManager? {
        return managers[blockchain.lowercase(Locale.US)]
            ?: managers.values.firstOrNull { it.spec.matches(blockchain) }
    }

    fun getWalletManagerBySpec(spec: BlockchainSpec): ChipWalletManager? =
        managers.values.firstOrNull { it.spec == spec }

    fun addWalletManagers(newManagers: List<ChipWalletManager>) {
        for (mgr in newManagers) {
            managers[mgr.spec.id] = mgr
            Log.d("ChipCoreNfc", "WalletManagerRegistry: added ${mgr.spec.id} address=${mgr.address}")
        }
    }

    fun clearWalletManagers() {
        Log.d("ChipCoreNfc", "WalletManagerRegistry: cleared ${managers.size} managers")
        managers.clear()
    }

    fun buildFrom(currencies: List<Messages.CurrencyInfoMessage>): List<ChipWalletManager> {
        return currencies.mapNotNull { currency ->
            val address = currency.address?.takeIf { it.isNotEmpty() } ?: return@mapNotNull null
            val publicKey = currency.publicKey ?: return@mapNotNull null
            val chainCode = currency.chainCode ?: byteArrayOf()
            val spec = BlockchainSpec.fromCurrency(currency)
            ChipWalletManager(
                spec = spec,
                address = address,
                publicKey = publicKey,
                chainCode = chainCode,
                isTest = currency.isTest == 1L,
            )
        }
    }

    fun all(): List<ChipWalletManager> = managers.values.toList()
}

// ─── CardCurrencyRepository ───────────────────────────────────────────────────

/**
 * 按卡片 UID 持久化币种列表（含已派生的地址和公钥），与 card_coin 的 CurrenciesRepository 对应。
 * 底层使用 SharedPreferences + JSON，无第三方依赖。
 */
internal class CardCurrencyRepository(private val activity: android.app.Activity) {
    private val prefs by lazy {
        activity.getSharedPreferences("chipcore_currencies", android.content.Context.MODE_PRIVATE)
    }

    /** 加载指定卡片的币种列表，未找到返回空列表 */
    fun load(cardId: String): List<Messages.CurrencyInfoMessage> {
        val json = prefs.getString("card_$cardId", null) ?: return emptyList()
        return try {
            val arr = org.json.JSONArray(json)
            (0 until arr.length()).mapNotNull { parseCurrency(arr.getJSONObject(it)) }
        } catch (_: Exception) {
            Log.w("ChipCoreNfc", "CardCurrencyRepository: failed to load currencies for $cardId")
            emptyList()
        }
    }

    /** 保存指定卡片的币种列表 */
    fun save(cardId: String, currencies: List<Messages.CurrencyInfoMessage>) {
        val arr = org.json.JSONArray()
        for (c in currencies) arr.put(encodeCurrency(c))
        prefs.edit().putString("card_$cardId", arr.toString()).apply()
        Log.d("ChipCoreNfc", "CardCurrencyRepository: saved ${currencies.size} currencies for $cardId")
    }

    private fun encodeCurrency(c: Messages.CurrencyInfoMessage): org.json.JSONObject {
        return org.json.JSONObject().apply {
            put("id", c.id)
            put("icon", c.icon ?: "")
            put("name", c.name)
            put("networkId", c.networkId)
            putOpt("networkName", c.networkName)
            put("networkIcon", c.networkIcon ?: "")
            put("symbol", c.symbol)
            putOpt("contractAddress", c.contractAddress)
            putOpt("decimalCount", c.decimalCount)
            putOpt("address", c.address)
            putOpt("publicKey", c.publicKey?.toHex())
            putOpt("chainCode", c.chainCode?.toHex())
            putOpt("isTest", c.isTest)
        }
    }

    private fun parseCurrency(o: org.json.JSONObject): Messages.CurrencyInfoMessage? {
        return try {
            val networkName = o.optString("networkName", "").takeIf { it.isNotEmpty() }
            val contractAddress = o.optString("contractAddress", "").takeIf { it.isNotEmpty() }
            val address = o.optString("address", "").takeIf { it.isNotEmpty() }
            val publicKeyHex = o.optString("publicKey", "").takeIf { it.isNotEmpty() }
            val chainCodeHex = o.optString("chainCode", "").takeIf { it.isNotEmpty() }
            Messages.CurrencyInfoMessage.Builder()
                .setId(o.getString("id"))
                .setIcon(o.optString("icon", ""))
                .setName(o.optString("name", ""))
                .setNetworkId(o.getString("networkId"))
                .setNetworkName(networkName)
                .setNetworkIcon(o.optString("networkIcon", ""))
                .setSymbol(o.optString("symbol", ""))
                .setContractAddress(contractAddress)
                .setDecimalCount(if (o.has("decimalCount")) o.getLong("decimalCount") else null)
                .setAmount(null) // 余额不持久化
                .setAddress(address)
                .setPublicKey(publicKeyHex?.hexToByteArrayOrNull())
                .setChainCode(chainCodeHex?.hexToByteArrayOrNull())
                .setIsTest(if (o.has("isTest")) o.getLong("isTest") else null)
                .build()
        } catch (e: Exception) {
            Log.w("ChipCoreNfc", "CardCurrencyRepository: failed to parse currency: ${e.message}")
            null
        }
    }
}

File: ios/Runner/WalletManager.swift
import Flutter
import Foundation

// MARK: - NativeWalletState（对应 Android WalletManager.kt）

final class NativeWalletState {
  var lastCardId: String?
  var lastPinSet: Bool = false
  private var currencies: [String: CurrencyInfoMessage] = [:]
  private var masterKeys: [String: KeyMaterial] = [:]

  func replaceCurrencies(_ newCurrencies: [CurrencyInfoMessage]) {
    // 清空前先快照已有的 publicKey/chainCode，避免 loadCurrencyInfoList 把
    // 卡片扫描存入的密钥数据覆盖掉
    let savedKeys: [String: CurrencyInfoMessage] = currencies.values.reduce(into: [:]) { dict, c in
      if c.publicKey != nil { dict[c.id] = c }
    }
    currencies.removeAll()
    for currency in newCurrencies {
      let toStore: CurrencyInfoMessage
      if currency.publicKey == nil, let saved = savedKeys[currency.id] {
        toStore = currency.copyWith(publicKey: saved.publicKey!.data, chainCode: saved.chainCode?.data)
      } else {
        toStore = currency
      }
      currencies[toStore.id] = toStore
      currencies[toStore.networkId] = toStore
    }
  }

  func allCurrencies() -> [CurrencyInfoMessage] {
    var seen = Set<String>()
    return currencies.values.filter { seen.insert($0.id).inserted }
  }

  func findById(_ id: String) -> CurrencyInfoMessage? {
    currencies[id]
  }

  func mergeCurrencies(_ newCurrencies: [CurrencyInfoMessage]) {
    for currency in newCurrencies {
      // 若新数据不含 publicKey/chainCode，保留已有条目的密钥信息（卡片扫描时写入）
      let existing = currencies[currency.id]
      let merged: CurrencyInfoMessage
      if currency.publicKey == nil, let existingKey = existing?.publicKey {
        merged = currency.copyWith(
          publicKey: existingKey.data,
          chainCode: existing?.chainCode?.data
        )
      } else {
        merged = currency
      }
      currencies[merged.id] = merged
      currencies[merged.networkId] = merged
    }
  }

  func removeCurrencies(_ coinIds: [String]) {
    for coinId in coinIds {
      currencies.removeValue(forKey: coinId)
    }
  }

  func findAddress(networkId: String) -> String? {
    currencies[networkId]?.address ?? currencies.values.first(where: { $0.networkId == networkId })?.address
  }

  func updateAddress(networkId: String, address: String) {
    guard let currency = currencies[networkId] ?? currencies.values.first(where: { $0.networkId == networkId }) else { return }
    let updated = currency.copyWith(address: address)
    currencies[currency.id] = updated
    currencies[currency.networkId] = updated
  }

  func findPublicKeyHex(keyword: String) -> String? {
    // 必须限定有 publicKey 的 currency，避免同一 networkId 下 ERC20 token
    // (无 publicKey) 被先迭代到，导致误返回 nil。
    currencies.values.first {
      $0.publicKey != nil && (
        $0.networkId.localizedCaseInsensitiveContains(keyword) ||
        $0.symbol.localizedCaseInsensitiveContains(keyword) ||
        $0.name.localizedCaseInsensitiveContains(keyword)
      )
    }?.publicKey?.data.hexString
  }

  func findBySpec(_ spec: BlockchainSpec) -> CurrencyInfoMessage? {
    currencies.values.first {
      spec.matches($0.networkId) || spec.matches($0.symbol) || spec.matches($0.name)
    }
  }

  func cacheMasterKey(cardId: String, publicKey: Data, chainCode: Data) {
    masterKeys[cardId] = KeyMaterial(publicKey: publicKey, chainCode: chainCode)
  }

  // MARK: - UserDefaults 持久化

  private static let storageKeyPrefix = "chip_wallet_currencies_"

  func saveCurrencies(cardId: String, currencies: [CurrencyInfoMessage]) {
    let key = Self.storageKeyPrefix + cardId
    let encoded = currencies.compactMap { currency -> [String: Any]? in
      var dict: [String: Any] = [
        "id": currency.id,
        "name": currency.name,
        "networkId": currency.networkId,
        "networkName": currency.networkName ?? "",
        "networkIcon": currency.networkIcon ?? "",
        "symbol": currency.symbol,
        "icon": currency.icon ?? "",
        "isTest": currency.isTest ?? 0,
      ]
      if let addr = currency.address { dict["address"] = addr }
      if let ca = currency.contractAddress { dict["contractAddress"] = ca }
      if let dc = currency.decimalCount { dict["decimalCount"] = dc }
      if let pk = currency.publicKey { dict["publicKey"] = pk.data.base64EncodedString() }
      if let cc = currency.chainCode { dict["chainCode"] = cc.data.base64EncodedString() }
      return dict
    }
    UserDefaults.standard.set(encoded, forKey: key)
  }

  func loadCurrencies(cardId: String) -> [CurrencyInfoMessage] {
    let key = Self.storageKeyPrefix + cardId
    guard let arr = UserDefaults.standard.array(forKey: key) as? [[String: Any]] else { return [] }
    return arr.compactMap { dict -> CurrencyInfoMessage? in
      guard let id = dict["id"] as? String,
            let name = dict["name"] as? String,
            let networkId = dict["networkId"] as? String,
            let symbol = dict["symbol"] as? String else { return nil }
      let msg = CurrencyInfoMessage(
        id: id,
        icon: dict["icon"] as? String ?? "",
        name: name,
        networkId: networkId,
        networkName: dict["networkName"] as? String,
        networkIcon: dict["networkIcon"] as? String ?? "",
        symbol: symbol,
        contractAddress: dict["contractAddress"] as? String,
        decimalCount: dict["decimalCount"] as? Int64,
        amount: nil,
        address: dict["address"] as? String,
        publicKey: (dict["publicKey"] as? String).flatMap { Data(base64Encoded: $0) }.map { FlutterStandardTypedData(bytes: $0) },
        chainCode: (dict["chainCode"] as? String).flatMap { Data(base64Encoded: $0) }.map { FlutterStandardTypedData(bytes: $0) },
        isTest: dict["isTest"] as? Int64
      )
      return msg
    }
  }

  func findContractAddress(symbol: String, networkId: String) -> String? {
    let sym = symbol.lowercased()
    let net = networkId.lowercased()
    return currencies.values.first {
      ($0.symbol.lowercased() == sym || $0.id.lowercased() == sym) &&
      $0.networkId.lowercased() == net &&
      ($0.contractAddress.map { !$0.isEmpty } ?? false)
    }?.contractAddress
  }

  func findDecimals(symbol: String, networkId: String) -> Int {
    let sym = symbol.lowercased()
    let net = networkId.lowercased()
    return currencies.values.first {
      ($0.symbol.lowercased() == sym || $0.id.lowercased() == sym) &&
      $0.networkId.lowercased() == net
    }.map { Int($0.decimalCount ?? 18) } ?? 18
  }
}

// MARK: - ChipWalletManager（对应 Android ChipWalletManager）

final class ChipWalletManager {
  let spec: BlockchainSpec
  let address: String
  let publicKey: Data
  let chainCode: Data
  let isTest: Bool
  var balance: String?

  init(spec: BlockchainSpec, address: String, publicKey: Data, chainCode: Data, isTest: Bool) {
    self.spec      = spec
    self.address   = address
    self.publicKey = publicKey
    self.chainCode = chainCode
    self.isTest    = isTest
  }
}

// MARK: - WalletManagerRegistry（对应 Android WalletManagerRegistry）

final class WalletManagerRegistry {
  private var managers: [String: ChipWalletManager] = [:]

  /// 通过 blockchain 字符串（如 "btc"、"ETH"）查找
  func getWalletManager(_ blockchain: String) -> ChipWalletManager? {
    managers[blockchain.lowercased()]
  }

  func getWalletManagerBySpec(_ spec: BlockchainSpec) -> ChipWalletManager? {
    managers.values.first { $0.spec == spec }
  }

  func addWalletManagers(_ newManagers: [ChipWalletManager]) {
    for manager in newManagers {
      managers[manager.spec.id.lowercased()] = manager
    }
  }

  func clearWalletManagers() {
    managers.removeAll()
  }

  /// 根据 CurrencyInfoMessage 列表构建 ChipWalletManager 列表
  func buildFrom(_ currencies: [CurrencyInfoMessage]) -> [ChipWalletManager] {
    currencies.compactMap { currency -> ChipWalletManager? in
      let spec     = BlockchainSpec.fromCurrency(currency)
      let isTest   = (currency.isTest ?? 0) != 0
      guard let address = currency.address, !address.isEmpty else { return nil }
      guard let pubKeyData = currency.publicKey?.data, !pubKeyData.isEmpty else { return nil }
      let chainCodeData = currency.chainCode?.data ?? Data()
      return ChipWalletManager(
        spec:      spec,
        address:   address,
        publicKey: pubKeyData,
        chainCode: chainCodeData,
        isTest:    isTest
      )
    }
  }

  func all() -> [ChipWalletManager] { Array(managers.values) }
}
